alacritty_terminal/src/term/text_run.rs

```diff
@@ -0,0 +1,309 @@
+        self.cell_at(self.span.1)
+    }
+
+    /// First point covered by this TextRun
+    pub fn start_point(&self) -> Point {
+        Point { line: self.line, col: self.span.0 }
+    }
+
+    /// End point covered by this TextRun
+    pub fn end_point(&self) -> Point {
+        Point { line: self.line, col: self.span.1 }
+    }
+
+    /// Returns iterator over range of columns [run.0, run.1]
+    pub fn cols(&self) -> impl Iterator<Item = Column> {
+        let (start, end) = self.span;
+        let step = if self.flags.contains(Flags::WIDE_CHAR) {
+            // If our run contains ide chars treat each cell like it's 2 cells wide
+            2
+        } else {
+            1
+        };
+        // unpacking is neccessary while Step trait is nightly
+        // hopefully this compiles away.
+        (start.0..=end.0).step_by(step).map(Column)
+    }
+
+    /// Iterates over each RenderableCell in column range [run.0, run.1]
+    pub fn cells<'a>(&'a self) -> impl Iterator<Item = RenderableCell> + 'a {
+        self.cols().map(move |col| self.cell_at(col))
+    }
+}
+
+/// Wraps an Iterator<Item=RenderableCell> and produces TextRuns to represent batches of cells
+pub struct TextRunIter<I> {
+    iter: I,
+    run_start: Option<RunStart>,
+    latest: Option<LatestCol>,
+    cursor: Option<CursorKey>,
+    buffer_text: String,
+    buffer_zero_width: Vec<[char; MAX_ZEROWIDTH_CHARS]>,
+}
+impl<I> TextRunIter<I> {
+    pub fn new(iter: I) -> Self {
+        TextRunIter {
+            iter,
+            latest: None,
+            run_start: None,
+            cursor: None,
+            buffer_text: String::new(),
+            buffer_zero_width: Vec::new(),
+        }
+    }
+
+    /// Check if current run ends with incoming RenderableCell
+    fn is_end_of_run(&self, rc: &RenderableCell) -> bool {
+        let is_cell_adjacent =
+            self.run_start.as_ref().map(|cell| !is_adjacent_context(cell, &rc)).unwrap_or(false);
+        let is_col_adjacent =
+            self.latest.as_ref().map(|col| !is_adjacent_col(*col, rc.column)).unwrap_or(false);
+        is_cell_adjacent || is_col_adjacent
+    }
+
+    /// Add content of cell to pending TextRun buffer
+    fn buffer_content(&mut self, inner: RenderableCellContent) {
+        // Add to buffer only if the next rc is a Char (not a cursor)
+        match inner {
+            RenderableCellContent::Chars(chars) => {
+                self.buffer_text.push(chars[0]);
+                let mut arr: [char; MAX_ZEROWIDTH_CHARS] = Default::default();
+                arr.copy_from_slice(&chars[1..]);
+                self.buffer_zero_width.push(arr);
+            },
+            RenderableCellContent::Cursor(cursor) => {
+                self.cursor = Some(cursor);
+            },
+        }
+    }
+
+    /// Empty out pending buffer producing owned collections that can be moved into a TextRun
+    fn drain_buffer(&mut self) -> (String, Vec<[char; MAX_ZEROWIDTH_CHARS]>) {
```

Comment:
This should be `bufferS`, it doesn't drain only one buffer. Documentation should state accordingly.

Reply:
Sure this is an implementation detail inside of text_run, but this method is private and only called inside of text_run. It's not hidden behind some kind of API that people aren't looking into. This is the place where things are implemented and it's returning two completely separate buffers not one. So I don't see how they are one.